won't make - not finding ncurses headers

I get the following error:

gcc -DSYSV3 -O2 -pipe -g   -c -o buffer.o buffer.c
In file included from buffer.c:3:0:
macros.h:47:21: fatal error: ncurses.h: No such file or directory
 #include <ncurses.h>
                     ^
compilation terminated.
make: *** [buffer.o] Error 1

It cannot seem to find <ncurses.h> which is installed.

Thanks for updating sc.

Phil hart

Sorry to write in an old issue, but I didn’t wanted to create new issue with the same problem.
OS: Arch Linux x86_64

Note: Actual question is at the end of this post.

#### What I’ve done
I just cloned `scim` to my computer:
```bash
git clone git@github.com:andmarti1424/sc-im.git scim
```
Then, I read the `Readme.md`, how I should install `scim`. It said that I should edit `Makefile` which I couldn’t _see_ any file called `Readme.md` (which is _usually_ located in base directory), so I searched for it:
```bash
find . | grep -i makefile
```
… so I found it in `$scim/src`. After I read it, I decided to leave it as is. In `$scim/src` directory, I ran:
```bash
make
```
and I got the following output:
```bash
gcc -DLINUX    -O2 -Wall -pipe -g -I/usr/include/libxml2 -DUSECOLORS -DUSELOCALE -DUNDO -DSIGVOID -DDFLT_PAGER=\"less\"   -DREGCOMP -DLIBDIR=\"/usr/share/doc/scim\" -DMAXROWS=65536 -DHELP_PATH=\"/usr/share/scim\" -DSNAME=\"scim\"     -DHISTORY_FILE=\".sciminfo\"   -c -o exec.o exec.c
In file included from exec.c:8:0:
macros.h:48:29: fatal error: ncursesw/curses.h: No such file or directory
compilation terminated.
<builtin>: recipe for target 'exec.o' failed
make: *** [exec.o] Error 1
```
At first, I checked if I have `ncurses` installed:
```bash
$ which ncurses
which: no ncurses in (/usr/lib:/usr/lib32:/root/.config/sublime-text-3/Packages:/home/ts/qt5.4.1/5.4/gcc/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games)

$ whereis ncurses
ncurses: /usr/include/ncurses.h /usr/share/man/man3/ncurses.3x.gz
```
… what is kind of strange (I couldn’t find, why `which` couldn’t find ncurses, but never mind). Then I realised (after reading some issues here on github.com for `scim`), that the `make` output was talking about `ncursesw`. So I checked again if I got i installed:
```bash
$ which ncursesw
which: no ncursesw in (/usr:/root/.config/sublime-text-3/Packages:/home/ts/qt5.4.1/5.4/gcc/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games)

$ whereis ncursesw
ncursesw:
```
… so I looked online how to install `ncursesw` on Arch linux. I found nothing, so I search for its installatino instructions on any linux, so I found out that is _should_ be isntalled by installing `ncurses`. So I search `/` for `ncurcesw`:
```bash
$ find / | grep ncursesw
/var/abs/community/newsbeuter/newsbeuter-2.9-ncursesw6.patch
/usr/bin/ncursesw6-config
/usr/lib/libncursesw.so
/usr/lib/libncursesw.so.6
/usr/lib/pkgconfig/ncursesw.pc
/usr/lib/libncursesw.so.6.0
/usr/lib32/libncursesw.so
/usr/lib32/libncursesw.so.6
/usr/lib32/libncursesw.so.6.0
^C
```
… indeed, it _is_ installed. My `$PATH` is as follow:
```bash
/usr:/root/.config/sublime-text-3/Packages:/home/ts/qt5.4.1/5.4/gcc/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
```

#### Actual question
The question is, why it is not working? I think that the `$scim/Readme.md` should include more detailed installation instructions (like where is the `Makefile`, where and how should `make` be executed, etc). Thank you in advanced

Thank you for your reply.

I found [this](https://bugs.archlinux.org/task/4418) site and then I realised that `ncursesw.h` is in Arch linux named as `cursesw.h`. So in `/usr/include` I’ve got `curses.h`, `cursesw.h` and a symlink to `curses.h` called `ncurses.h`. Creating such a symlink to curses.h didn’t help.

Then I edited `$scim/src/macro.h` file, name commented out 48th line and added new 49th line:

```bash
#include <cursesw.h>
```

… which solved the initial `ncursesw.h` error, but 13 new emerged with the final one being fatal. Here’s the output:

```bash
# make 
gcc -DLINUX    -O2 -Wall -pipe -g -I/usr/include/libxml2 -DUSECOLORS -DUSELOCALE -DUNDO -DSIGVOID -DDFLT_PAGER=\"less\"   -DREGCOMP -DLIBDIR=\"/usr/share/doc/scim\" -DMAXROWS=65536 -DHELP_PATH=\"/usr/share/scim\" -DSNAME=\"scim\"     -DHISTORY_FILE=\".sciminfo\"   -c -o exec.o exec.c
In file included from /usr/include/cursesw.h:36:0,
                 from macros.h:49,
                 from exec.c:8:
/usr/include/etip.h:118:8: error: expected identifier or ‘(’ before string constant
 extern "C" {
        ^
In file included from /usr/include/cursesw.h:36:0,
                 from macros.h:49,
                 from exec.c:8:
/usr/include/etip.h:142:1: error: unknown type name ‘class’
 class NCURSES_IMPEXP NCursesPanel;
 ^
/usr/include/etip.h:143:1: error: unknown type name ‘class’
 class NCURSES_IMPEXP NCursesMenu;
 ^
/usr/include/etip.h:144:1: error: unknown type name ‘class’
 class NCURSES_IMPEXP NCursesForm;
 ^
/usr/include/etip.h:146:1: error: unknown type name ‘class’
 class NCURSES_IMPEXP NCursesException
 ^
/usr/include/etip.h:147:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘{’ token
 {
 ^
/usr/include/etip.h:180:1: error: unknown type name ‘class’
 class NCURSES_IMPEXP NCursesPanelException : public NCursesException
 ^
/usr/include/etip.h:180:44: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘:’ token
 class NCURSES_IMPEXP NCursesPanelException : public NCursesException
                                            ^
/usr/include/etip.h:231:1: error: unknown type name ‘class’
 class NCURSES_IMPEXP NCursesMenuException : public NCursesException
 ^
/usr/include/etip.h:231:43: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘:’ token
 class NCURSES_IMPEXP NCursesMenuException : public NCursesException
                                           ^
/usr/include/etip.h:282:1: error: unknown type name ‘class’
 class NCURSES_IMPEXP NCursesFormException : public NCursesException
 ^
/usr/include/etip.h:282:43: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘:’ token
 class NCURSES_IMPEXP NCursesFormException : public NCursesException
                                           ^
/usr/include/etip.h:335:25: fatal error: iostream: No such file or directory
compilation terminated.
<builtin>: recipe for target 'exec.o' failed
make: *** [exec.o] Error 1
```

Indeed, mine is different. I tried to compile `scim` with your `curses.txt` (renamed to "curses_t.h") and the ncursesw.h error disappeared, but it needs `ncurses_dll.h`:
```bash
ncursesw/curses_t.h:60:34: fatal error: ncursesw/ncurses_dll.h: No such file or directory
```
Now, I’m trying to create the folder structure `scim` needs the `ncurses` files to be in (using your `curses.h` file located in /usr/include/ncurses/curses.h):
```bash
mkdir /usr/include/ncursesw
ln -s /usr/include/ncurses_dll.h /usr/include/ncursesw/ncurses_dll.h
ln -s /usr/include/unctrl.h /usr/include/ncursesw/unctrl.h
```
After doing this (and undoing every modification in `macro.h`), I’ve got this error:
```bash
gcc -DLINUX    -O2 -Wall -pipe -g -I/usr/include/libxml2 -DUSECOLORS -DUSELOCALE -DUNDO -DSIGVOID -DDFLT_PAGER=\"less\"   -DREGCOMP -DLIBDIR=\"/usr/share/doc/scim\" -DMAXROWS=65536 -DHELP_PATH=\"/usr/share/scim\" -DSNAME=\"scim\"     -DHISTORY_FILE=\".sciminfo\"   -c -o exec.o exec.c
In file included from macros.h:48:0,
                 from exec.c:8:
/usr/include/unctrl.h:60:38: error: conflicting types for ‘unctrl’
 NCURSES_EXPORT(NCURSES_CONST char *) NCURSES_SP_NAME(unctrl) (SCREEN*, chtype);
                                      ^
In file included from /usr/include/ncursesw/curses.h:2006:0,
                 from macros.h:48,
                 from exec.c:8:
/usr/include/unctrl.h:57:38: note: previous declaration of ‘unctrl’ was here
 NCURSES_EXPORT(NCURSES_CONST char *) unctrl (chtype);
                                      ^
<builtin>: recipe for target 'exec.o' failed
make: *** [exec.o] Error 1
```

--- a/src/Makefile
+++ b/src/Makefile
@@ -133,15 +133,16 @@ UNDO := -DUNDO
 # Can be set up to 1048576.
 MAXROWS := 65536
 
+CURSESFLAGS := $(shell pkg-config --cflags --libs ncursesw)
+
 #CFLAGS := -O2 -Wall -pipe -g
-CFLAGS := $(LINUX) $(FREEBSD) $(NETBSD) $(MACOSX) -O2 -Wall -pipe -g -I/usr/include/libxml2
+CFLAGS := $(LINUX) $(FREEBSD) $(NETBSD) $(MACOSX) -O2 -Wall -pipe -g -I/usr/include/libxml2 $(CURSESFLAGS)
 CFLAGS := $(CFLAGS) $(USECOLORS) $(USELOCALE) $(UNDO) $(SIGVOID) $(DFLT_PAGER)
 CFLAGS := $(CFLAGS) $(IEEE_MATH) $(RINT) $(REGEX) $(LIBRARY) -DMAXROWS=$(MAXROWS)
 CFLAGS := $(CFLAGS) $(HELP_PATH) $(SNAME) $(NO_NOTIMEOUT) $(SIMPLE) $(XLS) $(XLSX) $(HISTORY_FILE)
 
-# NOTE: ncursesw library should be used.
-# Not the ncurses library.
-LDLIBS := -lm -lncursesw
+
+LDLIBS := -lm
 #LDLIBS := -lm -lncurses -lxlsreader
 #LDLIBS := -lm -lncurses -lxlsreader -lzip -lxml2
 

--- a/src/cmds_command.c
+++ b/src/cmds_command.c
@@ -1,6 +1,6 @@
 #include <string.h>
 #include <ncurses.h>
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <wchar.h>
 #include <stdlib.h>
 #include <ctype.h>         // for isprint()

--- a/src/cmds_visual.c
+++ b/src/cmds_visual.c
@@ -1,4 +1,4 @@
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <stdlib.h>
 #include "utils/string.h"

--- a/src/color.c
+++ b/src/color.c
@@ -1,7 +1,7 @@
 #include <sys/types.h>
 #include <string.h>
 #include <stdlib.h>      // for atoi
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <ctype.h>
 #include <unistd.h>

--- a/src/file.c
+++ b/src/file.c
@@ -8,7 +8,7 @@
 #include <stdlib.h>
 #include <unistd.h>
 #include <wchar.h>
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <sys/wait.h>
 

--- a/src/help.c
+++ b/src/help.c
@@ -1,4 +1,4 @@
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <unistd.h>
 #include <stdlib.h>

--- a/src/history.c
+++ b/src/history.c
@@ -26,7 +26,7 @@ void set_comp(int i) {
 
 #if defined HISTORY_FILE
 
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <stdlib.h>
 #include <fcntl.h>

--- a/src/input.c
+++ b/src/input.c
@@ -1,7 +1,7 @@
 #include <sys/time.h>
 #include <string.h>
 #include <ctype.h>   // for isdigit
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <stdlib.h>
 #include <wchar.h>

--- a/src/interp.c
+++ b/src/interp.c
@@ -25,7 +25,7 @@
 #include <string.h>
 
 #include <stdlib.h>
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include "sc.h"
 #include "macros.h"

--- a/src/main.c
+++ b/src/main.c
@@ -1,5 +1,5 @@
 #include <signal.h>
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <unistd.h>
 #include <stdlib.h>

--- a/src/maps.c
+++ b/src/maps.c
@@ -1,6 +1,6 @@
 #include <stdlib.h>
 #include <ctype.h>
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <string.h>
 #include "maps.h"

--- a/src/range.c
+++ b/src/range.c
@@ -1,4 +1,4 @@
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <stdlib.h>
 #include "sc.h"

--- a/src/screen.c
+++ b/src/screen.c
@@ -7,7 +7,7 @@ main_win: window that loads the spreadsheetssword:
 
 
 #include <string.h>
-#include <ncursesw/curses.h>
+#include <curses.h>
 //#include <curses.h>
 #include <stdio.h>
 #include <stdlib.h>

--- a/src/sort.c
+++ b/src/sort.c
@@ -8,7 +8,7 @@ Adaptation of Chuck Martin's code - <nrocinu@myrealbox.com>
 #include <ctype.h>
 #include <stdlib.h>
 #include <unistd.h>
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include "macros.h"
 #include "yank.h"

--- a/src/vmtbl.c
+++ b/src/vmtbl.c
@@ -1,5 +1,5 @@
 #include <stdio.h>
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <unistd.h>
 #include "vmtbl.h"

--- a/src/xmalloc.c
+++ b/src/xmalloc.c
@@ -1,4 +1,4 @@
-//#include <ncursesw/curses.h>
+//#include <curses.h>
 #include <ncurses.h>
 #include <stdlib.h>
 #include "sc.h"

